src/main/scala/org/broadinstitute/dsde/workbench/sam/google/GoogleKeyCache.scala

```diff
@@ -0,0 +1,93 @@
+import com.google.api.services.storage.model.StorageObject
+import org.broadinstitute.dsde.workbench.google.{GoogleIamDAO, GoogleStorageDAO}
+import org.broadinstitute.dsde.workbench.model.google.GoogleModelJsonSupport._
+import org.broadinstitute.dsde.workbench.model.google.{GoogleProject, ServiceAccountKey, ServiceAccountKeyId}
+import org.broadinstitute.dsde.workbench.model._
+import org.broadinstitute.dsde.workbench.sam.config.{GoogleServicesConfig, PetServiceAccountConfig}
+import org.broadinstitute.dsde.workbench.sam.directory.DirectoryDAO
+import org.broadinstitute.dsde.workbench.sam.service.KeyCache
+
+import scala.concurrent.{ExecutionContext, Future}
+import spray.json._
+
+/**
+  * Created by mbemis on 1/10/18.
+  */
+class GoogleKeyCache(val directoryDAO: DirectoryDAO, val googleIamDAO: GoogleIamDAO, val googleStorageDAO: GoogleStorageDAO, val googleExtensions: GoogleExtensions, val googleServicesConfig: GoogleServicesConfig, val petServiceAccountConfig: PetServiceAccountConfig)(implicit val executionContext: ExecutionContext) extends KeyCache {
+
+  override def onBoot(): Future[Unit] = {
+    googleStorageDAO.createBucket(googleServicesConfig.serviceAccountClientProject, petServiceAccountConfig.keyBucketName).map { _ =>
+      googleStorageDAO.setBucketLifecycle(petServiceAccountConfig.keyBucketName, petServiceAccountConfig.activeKeyMaxAge)
+    }
+  }
+
+  private def keyNamePrefix(project: GoogleProject, saEmail: WorkbenchEmail) = s"${project.value}/${saEmail.value}"
+  private def keyNameFull(project: GoogleProject, saEmail: WorkbenchEmail, keyId: ServiceAccountKeyId) = s"${keyNamePrefix(project, saEmail)}/${keyId.value}"
+
+  def getKey(userEmail: WorkbenchEmail, project: GoogleProject): Future[Option[String]] = {
+    for {
+      subject <- directoryDAO.loadSubjectFromEmail(userEmail)
+      result <- subject match {
+        case Some(userId: WorkbenchUserId) => getKey(WorkbenchUser(userId, userEmail), project).map(Option(_))
+        case _ => Future.successful(None)
+      }
+    } yield result
+  }
+
+  private def createFreshKey(user: WorkbenchUser, project: GoogleProject): Future[String] = {
+    for {
+      pet <- googleExtensions.createUserPetServiceAccount(user, project)
+      key <- googleIamDAO.createServiceAccountKey(project, pet.serviceAccount.email) recover {
+        case e: GoogleJsonResponseException if e.getDetails.getCode == StatusCodes.TooManyRequests.intValue => throw new WorkbenchException("You have reached the 10 key limit on service accounts. Please remove one to create another.")
+      }
+      _ <- googleStorageDAO.storeObject(petServiceAccountConfig.keyBucketName, keyNameFull(project, pet.serviceAccount.email, key.id), new ByteArrayInputStream(key.privateKeyData.decode.get.getBytes))
+    } yield key.privateKeyData.decode.getOrElse(throw new WorkbenchException(s"could not decode service account private key data for key ${key.id}"))
```

Comment:
only decode once and this exception will never happen on this line because an exception would be thrown first on the line above